notes.md

```diff
@@ -0,0 +1,264 @@
+
+
+### Declaring a Listener
+First, define the listener in their app's `tests.listeners.py` file:
+
+```python
+post_reads_per_date_cache = defaultdict(dict)  # dummy cache for illustrative purposes only
+
+@listen('post_reads_per_date')
+def post_reads_per_date(
+        *,
+        post_id: int,
+        date: datetime.date,
+):
+    print(f'Processing post_reads_per_date with args {post_id}, {date}')
+    print(f'Cache before: {post_reads_per_date_cache}')
+    current_count = post_reads_per_date_cache[date].get(post_id, 0)
+    post_reads_per_date_cache[date][post_id] = current_count + 1
+    print(f'Cache after: {post_reads_per_date_cache}')
+```
+
+The user must also ensure that this `listeners` module is imported into
+that app's config class (similar to signals):
+```python
+class TestsConfig(AppConfig):
+    name = 'tests'
+
+    def ready(self):
+        import tests.listeners
+```
+
+#### Registering listeners
+When the project is loaded, the `listen` decorator is called. This
+executes this piece of logic:
+```python
+CustomPayloadChannel.register(callback, channel_name)
+```
+This creates an instance of CustomPayloadChannel and adds it to a global
+registry of "channels".
+
+##### Channels
+As in the postgres listen/notify protocol, messages are sent and received
+via named channels. The objects defined in channels.py are python abstractions
+of these channels. Instances of these classes have the following responsibilities:
+
+- Define an appropriate channel name if one has not been supplied.
+    - Undecided on how to best generate the name - see comments in _create_name.
+    - This is why in the examples I just define the names manually.
+- Listen to notifications on the channel matching their name.
+    - See the `listen` method
+- Deserialize notifications appropriately for the stored callback (the one
+  appearing under the `listen` decorator).
+    - I would really like the user to be able to notify the above channel with
+      the same payload as is accepted by the `post_reads_per_date` callback.
+    - I also want the callback to be able to accept a reasonable variety of different
+      python objects.
+    - So e.g. like `notify('post_reads_per_date', post_id=pk, date=datetime.date.today())`
+    - This means we need a way to easily serialize and then deserialize these arguments
+      back into python objects.
+    - The obvious choice for this is would be to use `pickle` as the serialization method. This
+      however did not work easily as the byte string pickling produced was not accepted by the
+      `pg_notify` function. Some sort of replace on the bad characters might let it, but that
+      felt a bit hacky.
+    - Instead, I've gone with JSON serialization. In order to be able to correctly deserialize
+      back to the same object types from a purely JSON payload, I have insisted that the
+      user define type annotations on their callback. These type annotations are then used
```

Comment:
I mentioned this possibility earlier, but I think a possibly friendlier user interface would be defining the payload schema for the channel itself. Is it valid to assume that a channel should have only one type of payload?

If a user writes many listeners for the same channel, I'm guessing it could be cumbersome to annotate the types on each listener each time. I'm assuming that's what's being proposed here.

Going to the dataclasses example one more time, assume you have code like this:

```
# In channels.py

@dataclass
class MyChannel(BaseChannel):
    post_id: int
    stuff: str
```

If the schema is defined on the channel, you give users the ability to either list out all keyword arguments they expected or to simply use a `**kwargs` for the ones they don't care about (similar to django signal handlers)

The code for serializing the datalclass should be as trivial as `json.dumps(MyChannel.as_dict())` too. Of course the serialization/de-serialization might be a little tricky with custom types as you mentioned